api/valor_api/backend/core/llm_clients.py

```diff
@@ -0,0 +1,521 @@
+
+    Evaluation Steps:
+    1. Read the source text carefully and identify the main topic and key points.
+    2. Check if the source text presents the information in a clear and logical order. Examine the collective quality of all sentences.
+    3. Assign a score for coherence on a scale of 1 to 5, where 1 is the lowest and 5 is the highest based on the Evaluation Criteria. Respond with just the number 1 to 5.
+
+    Source Text:
+    {text}
+
+    Coherence Score (1-5):
+    """
+
+
+class LLMClient:
+    """
+    Parent class for all LLM clients.
+
+    Attributes
+    ----------
+    api_key : str, optional
+        The API key to use.
+    model_name : str
+        The model to use.
+    """
+
+    api_key: str | None = None
+    model_name: str
+
+    def __init__(
+        self,
+        api_key: str | None = None,
+        model_name: str | None = None,
+    ):
+        """
+        Set the API key and model name (if provided).
+        """
+        self.api_key = api_key
+        if model_name is not None:
+            self.model_name = model_name
+
+    def connect(
+        self,
+    ):
+        """
+        Setup the connection to the API. Not implemented for parent class.
+        """
+        raise NotImplementedError
+
+    def process_messages(
+        self,
+        messages: list[dict[str, str]],
+    ) -> Any:
+        """
+        Format messages for the API.
+
+        Parameters
+        ----------
+        messages: list[dict[str, str]]
+            The messages formatted according to the OpenAI standard. Each message in messages is a dictionary with "role" and "content" keys.
+
+        Returns
+        -------
+        Any
+            The messages formatted for the API. By default, the messages are left in the OpenAI format.
+        """
+        return messages
+
+    def __call__(
+        self,
+        messages: list[dict[str, str]],
+    ) -> Any:
+        """
+        Call to the API. Not implemented for parent class.
+        """
+        raise NotImplementedError
+
+    def _generate_statements(
+        self,
+        text: str,
+    ) -> list[str]:
+        """
+        Generate a list of statements from a piece of text, using a call to the LLM API.
+
+        Parameters
+        ----------
+        text: str
+            The text to extract statements from.
+
+        Returns
+        -------
+        list[str]
+            The list of statements extracted from the text.
+        """
+        messages = [
+            {"role": "system", "content": DEFAULT_SYSTEM_PROMPT},
+            {
+                "role": "user",
+                "content": _generate_statements_instruction(text),
+            },
+        ]
+
+        response = self(messages)
+        try:
+            response = trim_and_load_json(response)
+            assert type(response) == dict
+            assert "statements" in response
+            statements = response["statements"]
+        except Exception:
```

Comment:
This looks messy and doesn't indicate to the user what the actual issue was with response. Also, the use of assert statements makes it unclear which fail conditions are actually being tested.